core/src/main/java/fr/sncf/osrd/envelope/EnvelopePhysics.java

```diff
@@ -174,4 +179,59 @@ public static double intersectStepWithSpeed(
             interpolatedPosition = minPos;
         return interpolatedPosition;
     }
+
+    /** Returns the cumulative energy consumed of the given envelope at the wheels of the train, given a certain path
+     * and rolling stock */
+    public static double getMechanicalEnergyConsumed(
+            Envelope envelope,
+            PhysicsPath path,
+            PhysicsRollingStock rollingStock
+    ) {
+        var cumulativeEnergy = 0.0;
+        for (var i = 0; i < envelope.size(); i++) {
+            var part = envelope.get(i);
+            cumulativeEnergy += getPartMechanicalEnergyConsumed(part, path, rollingStock);
+        }
+        return cumulativeEnergy;
+    }
+
+    /** Returns the total mechanical energy consumed on a given envelopePart */
+    public static double getPartMechanicalEnergyConsumed(
+            EnvelopePart part,
+            PhysicsPath path,
+            PhysicsRollingStock rollingStock
+    ) {
+        // The energy consumed by the train corresponds to the kinetic energy delta, subtracting the work by
+        // gravity and drag / friction
+        var length = part.pointCount();
+        var mass = rollingStock.getMass();
+        var inertia = rollingStock.getInertia();
+
+        var meanGrade = 0.001 * path.getAverageGrade(part.getBeginPos(), part.getEndPos());
+        var altitudeDelta = meanGrade * part.getTotalDistance();
+
+        var workGravity = - mass * 9.81 * altitudeDelta;
+
+        var kineticEnergyDelta =
+                0.5 * inertia * (part.getEndSpeed() * part.getEndSpeed() - part.getBeginSpeed() * part.getBeginSpeed());
+
+        var workDrag = 0;
+        for (var i = 0; i < length - 1; i++) {
+            var speed = part.getPointSpeed(i);
+            var nextSpeed = part.getPointSpeed(i + 1);
+            var meanSpeed = (speed + nextSpeed) / 2;
+            var pos = part.getPointPos(i);
+            var nextPos = part.getPointPos(i + 1);
+            var positionDelta = nextPos - pos;
+            workDrag -= rollingStock.getRollingResistance(meanSpeed) * positionDelta;
+        }
+
+        var totalEnergyConsumed = kineticEnergyDelta - workGravity - workDrag;
+
+        // If the train is braking the result should be negative
+        if (part.hasAttr(EnvelopeProfile.BRAKING))
+            assert totalEnergyConsumed <= 0;
```

Comment:
isn't this assert not always true? I call we encountered such a case while testing. it should also be 
```suggestion
assert !part.hasAttr(EnvelopeProfile.BRAKING) || totalEnergyConsumed <= 0;
```

Reply:
It's supposed to be, we just assert this for debugging purposes. Your suggestion makes sense though